### PR TITLE
[Deque] check deque equality with buffer identity

### PR DESCRIPTION
**Pull request by vanvoorden**
### Background
https://github.com/apple/swift-collections/pull/335

Similar to what we added for `OrderedDictionary`, we can try and optimize our `Deque` equality checks for the case where two `Deque` instances point to the same identity.

Unlike `OrderedDictionary`, our `Deque` is not implemented on `ContiguousArray`. We can try and mimic the approach from `ContiguousArray` here in `Deque` by checking against the identity of the underlying `ManagedBufferPointer` instance.

---

### Changes

From:
```
public static func ==(left: Self, right: Self) -> Bool {
  left.elementsEqual(right)
}
```
To:
```
let lhsCount = left.count
if lhsCount != right.count {
  return false
}

// Test referential equality.
if lhsCount == 0 || left._storage.isIdentical(to: right._storage) {
  return true
}

return left.elementsEqual(right)
```

---

### Test Plan

We leverage the existing `DequeTests`.

---

### Open Questions

Are there any additional edge-casey behaviors we should be testing for?

---

### Checklist
- [x] I've read the [Contribution Guidelines](/README.md#contributing-to-swift-collections)
- [x] My contributions are licensed under the [Swift license](/LICENSE.txt).
- [x] I've followed the coding style of the rest of the project.
- [x] I've added tests covering all new code paths my change adds to the project (if appropriate).
- [x] I've added benchmarks covering new functionality (if appropriate).
- [x] I've verified that my change does not break any existing tests or introduce unexplained benchmark regressions.
- [ ] I've updated the documentation if necessary.

### CODE REVIEW

**Comment by vanvoorden** [comment]
Here are benchmarks from running locally.

File: Tests/DequeTests/DequeTests.swift [hunk 1]
```diff
@@ -369,6 +369,23 @@ final class DequeTests: CollectionTestCase {
...
+    let deque: Deque<Int> = [1, 2, 3, 4]
+    let copy = deque
+    expectEqual(deque, copy)
+  }
+  
+  func test_not_equal() {
+    let left: Deque<Int> = [1, 2, 3, 4]
+    let right: Deque<Int> = [1, 2, 3, 4, 5]
+    expectNotEqual(left, right)
+  }
```
Comment (lorentey) [inline Tests/DequeTests/DequeTests.swift:383]: The `==` implementation is already covered in a more systematic way in the `test_Equatable_Hashable` test above. I think we can safely remove these.

```suggestion
```

File: Tests/DequeTests/DequeTests.swift [hunk 2]
```diff
@@ -369,6 +369,23 @@ final class DequeTests: CollectionTestCase {
...
+  
+  func test_equal_elements() {
+    let deque: Deque<Int> = [1, 2, 3, 4]
+    expectEqualElements(deque, [1, 2, 3, 4])
+  }
```
Comment (lorentey) [inline Tests/DequeTests/DequeTests.swift:388]: Similarly, the `test_CollectionConformance` test is already systematically checking that deques (of various shapes) do always produce the expected contents. (It also checks all possible methods of iterating through the deque, not just the one that `elementsEqual` uses.) This test can therefore be safely omitted.

```suggestion
```

File: Sources/DequeModule/Deque+Equatable.swift [hunk 3]
```diff
@@ -17,6 +17,16 @@ extension Deque: Equatable where Element: Equatable {
   /// - Complexity: O(`min(left.count, right.count)`)
   @inlinable
   public static func ==(left: Self, right: Self) -> Bool {
+    let lhsCount = left.count
+    if lhsCount != right.count {
+      return false
+    }
+
+    // Test referential equality.
+    if lhsCount == 0 || left._storage._buffer.buffer === right._storage._buffer.buffer {
```
Comment (lorentey) [inline Sources/DequeModule/Deque+Equatable.swift:26]: Super minor nit: We usually try to avoid directly accessing low-level layers within the implementation of high-level public API like this `==` implementation. It could be worth adding an `isIdentical(to:)` method to `Deque._Storage`, and then calling it here, instead of directly extracting the buffer references.

**Review by lorentey — APPROVED** [review]
Nice 👍

I picked some minor nits below.

**Comment by lorentey** [comment]
Note to self: letting `Deque.SubSequence` be the default `Slice<Deque>` means that it can't conform to `Equatable`/`Hashable`, and that is probably a defect in the Standard Library.

File: Sources/DequeModule/Deque._Storage.swift [hunk 4]
```diff
@@ -212,3 +212,11 @@ extension Deque._Storage {
     }
   }
 }
+
+extension Deque._Storage {
+  @inlinable
+  @inline(__always)
```
Comment (vanvoorden) [inline Sources/DequeModule/Deque._Storage.swift:218]: @lorentey Do these `inline` attributes look appropriate here?

Comment (lorentey) [reply Sources/DequeModule/Deque._Storage.swift:218]: This one looks fine to me! This package generally wants all generic code to be `@inlinable` to allow specialization (and to prevent having to materialize metatypes just to call this function). (This may not remain a tenable approach for long, but as long as these constructs are distributed as a package, the drawbacks of exposing function bodies are less pressing than e.g. if they shipped in the stdlib.)

`@inline(__always)` encourages the optimizer to actually inline the body more often into clients. In the case of `==`, I expect this doesn't matter much either way.

**Comment by lorentey** [comment]
@swift-ci test

**Review by lorentey — APPROVED** [review]
This looks ready to land! 🚢

**Comment by vanvoorden** [comment]
@lorentey Thanks for reviewing!
